### DISCUSSION TITLE
Diffuser Support for Invoke Users

### DISCUSSION THREAD

**Discussion by lottepeisch** [civitai/civitai#321] [Ideas]
InvokeAI has recently moved over to using diffusers instead of checkpoints. They can be bigger and take folders instead of a compressed checkpoint, but they load incredibly fast and Invoke connects to Huggingface to download them by repo. 

The only problem is with the rise of Civitai, fewer people are putting their models on Huggingface, and more and more are coming here. To use those models, I have to download and convert them, which would be fine if I were doing everything on my computer, but I use RunPod, so I end up having to download it from Civitai, convert it to diffuser format on the remote machine, then download it again to the host I'm using the next time I want to do it. 

It would be great to be able to skip that conversion step and pull down as a diffuser straight from Civit if that's possible, or maybe encourage a repo link to huggingface if that's not possible? Oh, and as long as that repo link is an actual weights repo, and not a checkpoint. There's a lot of folks who put checkpoints in both places and the weights are nowhere to be found. I'm less attached to the form of solution than I am that there is a solution. :)

**Comment by WASasquatch** [comment]
Don't support HF trying to reinvent a standard, is my opinion. Can't stand their mess over at diffusers. With ckpt and safetensors, there is no reason for diffusers, which uses filetypes known for issues, such as archiving, so storing them on online file hosts such as Google Drive, and then later downloading, is a pain. 

*PS I put a lot of time into development over at Diffusers since SD released for projects, and all they do is break everything downstream without notice or documentation. Additionally their philosophies are not conducive with the broader artistic field, such as a lot of civitai's focus.*

**Reply by lottepeisch** [reply]
Could you expand on that some more? I've been using diffusers for several weeks and haven't had any issues. In fact, I've been super pleased that I can switch models in a matter of seconds instead of minutes... what am I missing about their philosophies and/or the trouble with diffusers?

**Reply by WASasquatch** [reply]
> Could you expand on that some more? I've been using diffusers for several weeks and haven't had any issues. In fact, I've been super pleased that I can switch models in a matter of seconds instead of minutes... what am I missing about their philosophies and/or the trouble with diffusers?

As far as I am aware, it takes time to load models, and has been a main complaint of Diffusers is that it was actually slower then WebUI, unless they've finally implemented joblib or pickle caching of the whole object in a loaded state like I did for Easy Diffusion, which allowed loading models from cache in under 2 seconds, *after* they were cached out. 

But loading fast, is not something that is part of their models, but it's their API doing this. Their model format hasn't changed. 

As far as philosophies, there so reason for HF/Diffusers to try and impose a new format on a standard already used for years now except for it being proprietary to the Diffusers API. It's a means of locking down things, like they have done with safety checkers and such. They're just not conducive with Art as a whole, but their idea of it, as well as their idea of a format and use.

**Reply by lottepeisch** [reply]
> As far as I am aware, it takes time to load models, and has been a main complaint of Diffusers is that it was actually slower then WebUI, unless they've finally implemented joblib or pickle caching of the whole object in a loaded state like I did for Easy Diffusion, which allowed loading models from cache in under 2 seconds, _after_ they were cached out.

Oh no, InvokeAI's implementation of diffusers is leagues faster than WebUI. I'm too far removed from code to look through what they've done, but when you pop up a new repo into the models interface, it downloads the model, then as long as that's there, it takes 2-3 seconds to switch to that model, even if it's not one of the models cached in memory. It's ridiculously fast. Not sure if that's due to the API or not, though I suspect it's more Invoke doing it because even diffuser models not on HF load that fast. 

> As far as philosophies, there so reason for HF/Diffusers to try and impose a new format on a standard already used for years now except for it being proprietary to the Diffusers API. It's a means of locking down things, like they have done with safety checkers and such. They're just not conducive with Art as a whole, but their idea of it, as well as their idea of a format and use.

Hmm... no one's trying to impose anything, Invoke still has support for checkpoints and it doesn't look like they're going to pull that anytime soon. As for the reason for a different format, being able to change your scheduler and tokenizer and every other aspect of the model via Json config files, extremely fast load times, the HF API, and Invoke actually has a UX. The interface is clean, neat, and professional, and the canvas is AMAZING to work with. 

For example, I've been using 2dn for a while and I love it, but last night I decided to play around with different schedulers to see how it works, and with a few tweaks after some reading, I got a model I've had for a while to work even better and even faster than before, and when I made those same changes to one of my merges, it popped like crazy. And that was just changing the scheduler and tokenizer... 

But, uh... I don't see choosing to properly support MORE formats as a philosophical issue, it's a practicality issue. I'd counter your earlier statement that CivitAI doesn't want to stifle creativity by stating that ignoring one format isn't conducive to a creative artistic environment, rather it stifles it by alienating those that don't use the dominant format.

**Reply by bghira** [reply]
@WASasquatch can you please simply delete all of your misinformed comments on this discussion page? you are coming here and saying things without evidence and then insisting that you are correct when people push back.

**Comment by WASasquatch** [comment]
Maybe you're a little confused. 

InvokeAI is just using HF Diffusers API. Diffusers API has no native support for CKPT. It would be adding an entire framework, that's not compatible with WebUI, and broader AI tools. HF already has their own model community and ability to search, etc. They should just improve that. 

It's cool invokeAI is loading the diffusers models so fast, this is something I don't see with the API myself. But I suspect they are just caching out the model after downloading it.

There is already plenty of confusion with just uploading a Diffusers based embedding in `.bin` format. Lol Reason being is Diffusers API is a n API. It's meant for other developers to use in their services, where as what everyone is using at home, etc, is WebUI, using AI standards already years established,

Also to really nail the point in: Diffusers models are not friendly with file sharing services. You can't just archive up the model with most services, it will archive part of it, and store the largest bin separately, also renamed, which then requires users to rename and move files after download. That's just a hassle. That's why HuggingFace Hub is used to download the files programmatically with Python sequentially. It was a mistake for HuggingFace to use binary filetype, which is inherently a questionable format security wise, but also *meant* to be binary, thus archives have issues with it, or respect it as binary, and don't even archive it.

**Reply by bghira** [reply]
none of that is accurate, it's loaded with personal opinions, and there's no technical reason that it can't be done.

if you don't want to write the code, that's okay. but don't keep lying about the downsides so that others are discouraged from trying, either.

you have a bone to pick with huggingface? go do it on their discord and leave people here be alone.

**Reply by WASasquatch** [reply]
How about you explain why things are lies and how "none of it is accurate" instead of just saying they are, hmm? Most of this is common at a glance info of the repos, and yes HF is trying to push a new standard with Diffusers against already established standards in AI art communities. That's inherent.  That's no secret. And lots of people have beef with HF, including that. Check out their issues page and discussion. They do not listen to their users. 

HF has already stated he doesn't want to include diffusers into ComfyUI, to begin with. Adding diffusers converter to load a diffusers model to s standard format to use is what we got. 

And I can have opinions regarding an opinion of an idea. Simple as that.

**Reply by AkoZ** [reply]
it seems at start that comfy has diffusers .. ;) https://blenderneko.github.io/ComfyUI-docs/Core%20Nodes/Advanced/DIffusersLoader/ but then some can describe the process .. as in invokeai , when i've tested that it _was very fast to get a model in diffusers_ but it was a temp thing so.. it takes a time, relatively fast to convert..
Then for sure any other as webui or easydiff are so very very slow to pass from one model to an other. it needs a real plan to get the new touch, as to merge some parts to get the best merged-models-MODEL suitable for each.
* so .. ways now is to compete and ask if diffusers are a good thing or (as said, a security falling bridge) let it pass the way and don't care to get the most of models.safetensors being LORAified ?
* ok i'll stay there and give some search to have a global answer .. let's explore invokeai reasons. 
At start is : https://invoke-ai.github.io/InvokeAI/installation/050_INSTALLING_MODELS/#installation-via-the-web-gui
there about the diffusers, which seem to be downloadable by HF ? :

> InvokeAI supports all three formats but **strongly prefers the diffusers format**. 
> These are distributed as directories containing multiple subfolders, each of which contains a different aspect of the model.
> The advantage of this is that the models load from disk really fast.
> Another advantage is that diffusers models are supported by a large and active set of open source developers working at and with HuggingFace organization, and improvements in both rendering quality and performance are being made at a rapid pace. Among other features is the ability to download and install a diffusers model just by providing its HuggingFace repository ID. 

* then it appears a smart way to use the big "zipper" files checkpointsLIKE in their inside directories ?
* ok there are some diffusers in HF but not so easy to understand nor to use as they don't ref any image or usage ??.: https://huggingface.co/models?sort=trending&search=diffusers
* the thing is "[CIVITAI](https://civitai.com/). This extensive site currently supports only .safetensors and .ckpt models, **but they can be easily loaded into InvokeAI and/or converted into optimized diffusers models.**"
* 
![image](https://github.com/civitai/civitai/assets/449844/813a2b13-e16d-43ba-a7de-de4319867f98)
*
so diffusers in COMFY are deprecated ! .. well ok no advantages or something wrong with ? who else ? HF interest conflicting as nearly said in comments ? security breach (as it seems not if you DL your .safetensors and convert locally  ?

well tested diffusers were to be far faster than .safetensors to load and get an image with...
Please if someone can clarify the real issue with those appearing fast loading diffusers  ?

**Reply by WASasquatch** [reply]
I have a diffusers loader in WAS Node Suite that takes a repo ID from HF. comfyanonymous for some reason pushed hard about not including any autodownloads in base comfyui. That's why it expects an already downloaded model. Which imo is counterproductive, confusing, and may as well not been added to base ComfyUi
